subsys/bluetooth/audio/bap_base.c

```diff
@@ -0,0 +1,572 @@
+#define BASE_CODEC_ID_SIZE       (1 /* id */ + 2 /* cid */ + 2 /* vid */)
+#define BASE_PD_SIZE             3
+#define BASE_SUBGROUP_COUNT_SIZE 1
+#define BASE_NUM_BIS_SIZE        1
+#define BASE_CC_LEN_SIZE         1
+#define BASE_META_LEN_SIZE       1
+#define BASE_BIS_INDEX_SIZE      1
+#define BASE_BIS_CC_LEN_SIZE     1
+#define BASE_SUBGROUP_MAX_SIZE   (BASE_MAX_SIZE - BASE_PD_SIZE - BASE_SUBGROUP_COUNT_SIZE)
+#define BASE_SUBGROUP_MIN_SIZE                                                                     \
+	(BASE_NUM_BIS_SIZE + BASE_CODEC_ID_SIZE + BASE_CC_LEN_SIZE + BASE_META_LEN_SIZE +          \
+	 BASE_BIS_INDEX_SIZE + BASE_BIS_CC_LEN_SIZE)
+#define BASE_MIN_SIZE                                                                              \
+	(BT_UUID_SIZE_16 + BASE_PD_SIZE + BASE_SUBGROUP_COUNT_SIZE + BASE_SUBGROUP_MIN_SIZE)
+#define BASE_SUBGROUP_MAX_COUNT (BASE_MAX_SIZE / BASE_SUBGROUP_MIN_SIZE)
+
+static uint32_t base_pull_pd(struct net_buf_simple *net_buf)
+{
+	return net_buf_simple_pull_le24(net_buf);
+}
+
+static uint8_t base_pull_bis_count(struct net_buf_simple *net_buf)
+{
+	return net_buf_simple_pull_u8(net_buf);
+}
+
+static void base_pull_codec_id(struct net_buf_simple *net_buf,
+			       struct bt_bap_base_codec_id *codec_id)
+{
+	struct bt_bap_base_codec_id codec;
+
+	codec.id = net_buf_simple_pull_u8(net_buf);    /* coding format */
+	codec.cid = net_buf_simple_pull_le16(net_buf); /* company id */
+	codec.vid = net_buf_simple_pull_le16(net_buf); /* VS codec id */
+
+	if (codec_id != NULL) {
+		*codec_id = codec;
+	}
+}
+
+static uint8_t base_pull_ltv(struct net_buf_simple *net_buf, uint8_t **data)
+{
+	const uint8_t len = net_buf_simple_pull_u8(net_buf);
+
+	if (data == NULL) {
+		net_buf_simple_pull_mem(net_buf, len);
+	} else {
+		*data = net_buf_simple_pull_mem(net_buf, len);
+	}
+
+	return len;
+}
+
+const struct bt_bap_base *bt_bap_base_get_base_from_ad(const struct bt_data *ad)
+{
+	struct bt_uuid_16 broadcast_uuid;
+	const struct bt_bap_base *base;
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+	void *uuid;
+
+	CHECKIF(ad == NULL) {
+		LOG_DBG("data is NULL");
+
+		return NULL;
+	}
+
+	if (ad->type != BT_DATA_SVC_DATA16) {
+		LOG_DBG("Invalid type: %u", ad->type);
+
+		return NULL;
+	}
+
+	if (ad->data_len < BASE_MIN_SIZE) {
+		LOG_DBG("Invalid len: %u", ad->data_len);
+
+		return NULL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)ad->data, ad->data_len);
+
+	uuid = net_buf_simple_pull_mem(&net_buf, BT_UUID_SIZE_16);
+	if (!bt_uuid_create(&broadcast_uuid.uuid, uuid, BT_UUID_SIZE_16)) {
+		LOG_ERR("bt_uuid_create failed");
+
+		return NULL;
+	}
+
+	if (bt_uuid_cmp(&broadcast_uuid.uuid, BT_UUID_BASIC_AUDIO) != 0) {
+		LOG_DBG("Invalid UUID");
+
+		return NULL;
+	}
+
+	/* Store the start of the BASE */
+	base = (const struct bt_bap_base *)net_buf.data;
+
+	/* Pull all data to verify that the result BASE is valid */
+	base_pull_pd(&net_buf);
```

Comment:
A reference to the BAP spec, table 3.14 would be nice IMO

Reply:
Further up in the file we have 

```c
/* The BASE and the following defines are defined by BAP v1.0.1, section 3.7.2.2 Basic Audio
 * Announcements
 */
```

Is that good enough? It does not make much sense to add a reference for each value in the BASE